updating developer documentation (#442)
GoogleCloudDataproc/spark-bigquery-connector: CONTRIBUTING.md

## Building and Testing the Connector

The connector is built using the Maven wrapper. The project contains several
connectors, sharing some of the code among them. The connectors are:
* spark-bigquery_2.11 - a Scala 2.11 based connector, targeting Spark 2.3 and
  2.4 using Scala 2.11.
* spark-bigquery_2.12 - a Scala 2.12 based connector, targeting Spark 2.4 and 
  3.x using Scala 2.12.
* spark-bigquery:spark2.4 - a Java only connector, targeting Spark 2.4 (of all
  Scala versions), using the new DataSource APIs.
* spark-bigquery:spark3 - a Java only connector, targeting Spark 3 (of all
  Scala versions), using the new DataSource APIs. Still under development.

The project's artifacts are:
  
* `spark-bigquery-parent` - The parent POM for all artifacts. Common settings
  and artifact version should be defined here.
* `bigquery-connector-common` - Utility classes for working with the BigQuery
  APIs. This artifact has no dependency on Spark.
* `spark-bigquery-connector-common`- Common code shared among all connectors

>   (Scala and Java alike).

Reviewer: `bigquery-connector-common`  v/s  `spark-bigquery-connector-common` looks a bit confusing. Maybe we should add a line or two.

  distributable of the connector, containing all the dependencies.
* `spark-bigquery-dsv2/spark-bigquery-dsv2-parent` - Common settings for the
  Java only DataSource V2 implementations.
* `spark-bigquery-dsv2/spark-bigquery-spark24` - A Java only DataSource V2
  connector implementing the Spark 2.4 APIs.
* `spark-bigquery-dsv2/spark-bigquery-spark3` - A Java only DataSource V2
  connector implementing the Spark 3 APIs. Under development.
* `spark-bigquery-python-lib` - The python support library, adding BigQuery
  types not supported by Spark.

As building and running all the connectors is a lengthy process, the project is
split into several artifacts, and only part of them can be built at a single run
using [Maven profiles](https://maven.apache.org/guides/introduction/introduction-to-profiles.html).
The profiles are:

* `dsv1` - Running both Scala/DSv1 connectors.
* `dsv1_2.11` - Running just the Scala 2.11 connector.
* `dsv1_2.12` - Running just the Scala 2.12 connector.
* `dsv2` - Running both Java/DSv2 connectors.
* `dsv2_2.4` - Running just the Java Spark 2.4 connector.
* `dsv2_3` - Running just the Java Spark 3 connector.
* `all` - Running all the connectors.

> 

Reviewer: We can add a small example here, how to run these profiles, how to build the connector or how to create the jar. Similar to the one given below.

* `spark-bigquery-dsv1/spark-bigquery_2.11` and
  `spark-bigquery-dsv1/spark-bigquery_2.12` - The implementation of the Scala
  based connectors. Both connectors share the same code via a symbolic link, so
  a change in one of them will automatically affect the other.
* `spark-bigquery-dsv1/spark-bigquery-with-dependencies-parent` - Common
  settings for the shaded artifacts.
* `spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.11` and
  `spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.12` - The shaded
  distributable of the connector, containing all the dependencies.
* `spark-bigquery-dsv2/spark-bigquery-dsv2-parent` - Common settings for the
  Java only DataSource V2 implementations.
* `spark-bigquery-dsv2/spark-bigquery-spark24` - A Java only DataSource V2
  connector implementing the Spark 2.4 APIs.
* `spark-bigquery-dsv2/spark-bigquery-spark3` - A Java only DataSource V2
  connector implementing the Spark 3 APIs. Under development.
* `spark-bigquery-python-lib` - The python support library, adding BigQuery
  types not supported by Spark.

As building and running all the connectors is a lengthy process, the project is

> split into several artifacts, and only part of them can be built at a single run

Reviewer 2: ```suggestion
split into several artifacts. Each can be selectively built
```

using [Maven profiles](https://maven.apache.org/guides/introduction/introduction-to-profiles.html).

> The profiles are:

Reviewer 2: the preceding paragraph mentions 'artifacts' not 'profiles' should there be consistency in the language?

Author: A profile refers to several artifacts, for example `dsv1_2.12` -> `spark-bigquery-dsv1-spark3-support`, `spark-bigquery-dsv1-parent`,`spark-bigquery_2.12`, `spark-bigquery-with-dependencies-parent`, `spark-bigquery-with-dependencies_2.12`

Author: I've re-phrased the paragraph, PTAL
